(DOCSP-19708) Configuration (#4)
mongodb/docs-visual-studio-extension: source/configuration.txt

.. _visual-studio-extension-configuration:

==============================
Configure the {+product+}
==============================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

Learn how to configure the {+product+} in your .NET project and what
configuration options are available.

Configuration
-------------

To configure the {+product+}, perform the following actions:

> #. Install the {+product+} as described in the :doc:`Install </install>` guide.

Author: Responding in order:

- Agree ref linking is better. I'll add "change :doc: links to :ref:" as a [clean up task ](https://jira.mongodb.org/browse/DOCSP-20099) and update this to a "`<New text> <TODO: update to ref linking to install guide>`__.
- Based off of the [Links](https://docs.mongodb.com/meta/style-guide/style/links-and-cross-references/) entry in the style guide and your suggested change, I believe this link text should be updated to "the installation guide".

> When links are inline, use about three or four words of existing text as the link text. Choose words that best describe the target content.

#. Create a file specifying your configuration for the {+product+} named
   ``{+product-settings-file-constant+}``.
   
   .. important:: Naming

      You can put your settings file in any directory, but you must name your

>       file ``{+product-settings-file-constant+}``.

Author: I believe you specify the path to your config file later in the `AdditionalFiles` tag, but I agree saying "hey guess what you can put this file in a weird place" is not helpful for readers. I'll trim this down to "You must name your configuration file ``mongodb.analyzer.json``.

#. Add your configuration to your ``{+product-settings-file-constant+}`` file.

   The following sample configuration makes the {+product+} use the 
   `LINQ3 <{+driver-docs+}reference/driver/crud/linq3/>`__
   provider.

   .. code-block:: json

      { "DefaultLinqVersion": "V3" }

#. Add the following to the ``.csproj`` for your .NET application within the ``Project`` tag:

   .. code-block:: xml

      <ItemGroup>
        <AdditionalFiles Include="<path to your {+product-settings-file-constant+} file>" />
      </ItemGroup>

To learn more about .NET project files, see 
`Understanding the Project File <https://docs.microsoft.com/en-us/aspnet/web-forms/overview/deployment/web-deployment-in-the-enterprise/understanding-the-project-file>`__
from Microsoft.

To learn more about the ``AdditionalFiles`` tag, see 
:github:`Using Additional Files </dotnet/roslyn/blob/main/docs/analyzers/Using%20Additional%20Files.md>`
in the Roslyn documentation.

Options
-------

The following is a description of all configuration options supported
by the {+product+}:

.. list-table::
   :header-rows: 1
   :widths: 35 65

   * - Name
     - Description

   * - | **DefaultLinqVersion**

>      - | **Type:** string

Reviewer: question: Is type necessary since we're showing what the default values are?

Author: I got this [style of options from the Kafka Docs](https://docs.mongodb.com/kafka-connector/current/sink-connector/configuration-properties/mongodb-connection/), and I thought it was helpful there (I would rather explicitly state the type than have people potentially unfamiliar with this table's format infer it).

The only semi-concrete reason to keep type information here is to distinguish between boolean `true` and the string `"true"`.

I'm a fan of keeping the types (I think it might help someone understand it) but there is definitely an argument to remove them. If there is a strong opinion to remove them I can remove them, but I have a slight preference for keeping them.

Reviewer: I'd measure whether any perceived value gained correlates with measurable feedback from users. I'll leave it up to you.

       | Specifies if the {+product+} includes your {+driver-short+} version in diagnostic messages.
       |
       | **Default**: ``false``
       | **Accepted Values**: ``true`` or ``false``

   * - | **OutputInternalExceptions**
     - | **Type:** boolean
       |
       | **Description:**
       | Specifies if the {+product+} includes internal exceptions in diagnostic messages.
       |
       | **Default**: ``false``
       | **Accepted Values**: ``true`` or ``false``

   * - | **OutputInternalLogsToFile**
     - | **Type:** boolean
       |
       | **Description:**
       | Specifies if the {+product+} writes its internal logs to a file.

       .. include:: includes/logging-admonition.rst       

       | **Default**: ``false``
       | **Accepted Values**: ``true`` or ``false``

   * - | **LogFileName**
     - | **Type:** string
       |
       | **Description:**
       | Specifies the path to which the {+product+} writes its internal logs.
       
       .. include:: includes/logging-admonition.rst       

       | **Default**: ``""``
       | **Accepted Values**: A valid file path
   * - | **SendTelemetry**
     - | **Type:** boolean
       |
       | **Description:**

>        | Specifies if the {+product+} sends anonymized information to MongoDB (the company).

Reviewer: question: Why add "(the company)"? I can't find any other usage like this anywhere in the documentation. Could this be worded to avoid the confusion I'm guessing you're trying to avoid?

```suggestion
       | Specifies if the {+product+} sends anonymized information for continued improvement.
```

Author: The only other telemetry statement I was aware of was this [mongosh page](https://docs.mongodb.com/mongodb-shell/telemetry/) page Max directed me to, and I believe a shortcoming of this page is it does not explicitly answer the question "Who is getting my data?" instead only answering "Why are you collecting my data?" (Answer: "to improve products").

I think "(the company)" should be changed. I think "MongoDB Inc." might be a good alternative (this is used in the [Privacy Policy](https://www.mongodb.com/legal/privacy-policy)).

Reviewer: That works.

         To learn more, see the :ref:`telemetry <mongodb-analyzer-configuration-telemetry>`
         section of this guide.

       | **Default**: ``true``
       | **Accepted Values**: ``true`` or ``false``

To learn more about the configuration options the {+product+} supports,
see the :github:`{+product+} source code </master/src/MongoDB.Analyzer/Core/Settings/SettingsHelper.cs>`.

.. _mongodb-analyzer-configuration-telemetry:

Telemetry
~~~~~~~~~

> When the {+product+} analyzes your source code, the tool by default

Reviewer: question: same here about "(the company)" - would stating the why and removing MongoDB avoid this? I think most users just need to know that data is being sent and what it's for.

Author: I think "who is my data sent to?" is an important question to answer as well as "why is my data sent?". As stated in preceding response, I don't have a great citation for why this is important. If people on the team feel strongly that this question should not be answered here then I agree that "MongoDB Inc./MongoDB (the company)" does not have to be mentioned.

Reviewer: My main point was that MongoDB (the company) was breaking from convention.

.. _visual-studio-extension-configuration:

==============================
Configure the {+product+}
==============================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

Learn how to configure the {+product+} in your .NET project and what
configuration options are available.

Configuration
-------------

> To configure the {+product+}, perform the following actions:

Reviewer 2: Can we somehow emphasize that providing the config file is optional? Maybe something like "To override analyzer's default settings..."

Author: Good point, I agree that specifying that "configuration is optional" is beneficial to readers.

I think adding "to override the MongoDB Analyzer defaults...", while correct, may not directly engage with the task readers come to this page to learn how to perform "configure the MongoDB Analyzer".

I think placing an admonition stating "Configuration is Optional" after the steps to configure the MongoDB Analyzer may add this information while allowing the first sentence to keep the "To configure..." language.

Let me know if this fix doesn't address your concerns!
